components/datetime/src/date.rs

```diff
@@ -0,0 +1,245 @@
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::ops::{Add, Sub};
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub enum DateTimeError {
+    Parse(std::num::ParseIntError),
+    Overflow { field: &'static str, max: usize },
+}
+
+impl fmt::Display for DateTimeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Parse(err) => write!(f, "{}", err),
+            Self::Overflow { field, max } => write!(f, "{} must be between 0-{}", field, max),
+        }
+    }
+}
+
+impl From<std::num::ParseIntError> for DateTimeError {
+    fn from(input: std::num::ParseIntError) -> Self {
+        Self::Parse(input)
+    }
+}
+
+/// Temporary trait used to represent the input data for `DateTimeFormat`.
+///
+/// This type represents all data that the formatted needs in order to produced formatted string.
+///
+/// *Note*: At the moment we support only `gregorian` calendar, and plan to extend support to
+/// other calendars in the upcoming releases.
+pub trait DateTimeType: FromStr {
+    fn year(&self) -> usize;
+    fn month(&self) -> Month;
+    fn day(&self) -> Day;
+    fn hour(&self) -> Hour;
+    fn minute(&self) -> Minute;
+    fn second(&self) -> Second;
+}
+
+/// Temporary implementation of `DateTimeType`,
+/// which is used in tests, benchmarks and examples of this component.
+///
+/// # Examples
+///
+/// ```
+/// use icu_datetime::DummyDateTime;
+///
+/// let dt = DummyDateTime::try_new(2020, 9, 24, 13, 21, 0)
+///     .expect("Failed to construct DateTime.");
+/// ```
+#[derive(Debug, Default)]
+pub struct DummyDateTime {
+    pub year: usize,
+    pub month: Month,
+    pub day: Day,
+    pub hour: Hour,
+    pub minute: Minute,
+    pub second: Second,
+}
+
+impl DummyDateTime {
+    pub fn new(
+        year: usize,
+        month: Month,
+        day: Day,
+        hour: Hour,
+        minute: Minute,
+        second: Second,
```

Comment:
Nit: I think you added millisecond at some point, but now it's gone?

Reply:
Yes, sorry for not explanining. I initially prototyped a larger version of this api to understand the facets that included most of ldml, skeletons, components etc. Then, in result of our agreement to trim it to minimum I removed almost all pieces that weren't strictly needed and yet the review took over 150 comments. Shane called out a number of places where I bits and pieces were left that aren't needed for this PR and I went through a cleanse to remove them all. We don't use milliseconds in styles so I removed it and will bring back when we work on skeletons.
Is that ok with you?